lib/paloaltofw.py

```diff
@@ -0,0 +1,534 @@
+    name = address.parent_token
+    for ip in self.addressbook[zone][name]:
+      if str(address) == str(ip[0]):
+        return
+    counter = len(self.addressbook[zone][address.parent_token])
+    name = '%s_%s' % (name, str(counter))
+    self.addressbook[zone][address.parent_token].append((address, name))
+
+  def _BuildPort(self, ports):
+    """Transform specified ports into list and ranges.
+
+    Args:
+      ports: a policy terms list of ports
+
+    Returns:
+      port_list: list of ports and port ranges
+    """
+    port_list = []
+    for i in ports:
+      if i[0] == i[1]:
+        port_list.append(str(i[0]))
+      else:
+        port_list.append('%s-%s' % (str(i[0]), str(i[1])))
+    return port_list
+
+  def __str__(self):
+    """Render the output of the PaloAltoFirewall policy into config."""
+    initial = []
+    # INITAL CONFIG
+    initial.append('<?xml version="1.0"?>"')
+    initial.append('<config version="7.0.0" urldb="paloaltonetworks">')
+    initial.append(self.INDENT * 1 + '<devices>')
+    initial.append(self.INDENT * 2 + '<entry name="localhost.localdomain">')
+    initial.append(self.INDENT * 3 + '<vsys>')
+    initial.append(self.INDENT * 4 + '<entry name="vsys1">')
+    initial.append(self.INDENT * 5 + '<application/>')
+    initial.append(self.INDENT * 5 + '<application-group/>')
+
+    # SERVICES
+    service = []
+    service.append(self.INDENT * 5 + "<!-- Services -->")
+
+    service.append(self.INDENT * 5 + "<service>")
+    for k,v in Service.service_map.items():
+      service.append(self.INDENT * 6 + '<entry name="' + v["name"] + '">')
+      service.append(self.INDENT * 7 + "<protocol>")
+      service.append(self.INDENT * 8 + "<" + k[1] + ">")
+      tup = str(k[0])[1:-1]
+      if tup[-1] == ",":
+        tup = tup[:-1]
+      service.append(self.INDENT * 9 + "<port>" + tup.replace("'","") + "</port>")
+      service.append(self.INDENT * 8 + "</" + k[1] + ">")
+      service.append(self.INDENT * 7 + "</protocol>")
+      service.append(self.INDENT * 6 + "</entry>")
+    service.append(self.INDENT * 5 + "</service>")
+
+    # RULES
+    rules = []
+    rules.append(self.INDENT * 5 + "<!-- Rules -->")
+
+    rules.append(self.INDENT * 5 + "<rulebase>")
+    rules.append(self.INDENT * 6 + "<security>")
+    rules.append(self.INDENT * 7 + "<rules>")
+
+    for name, options in Rule.rules.items():
+      rules.append(self.INDENT * 8 + '<entry name="' + name + '">')
+
+      rules.append(self.INDENT * 9 + "<to>")
+      for tz in options["to_zone"]:
+        rules.append(self.INDENT * 10 + "<member>" + tz + "</member>")
+      rules.append(self.INDENT * 9 + "</to>")
+
+      rules.append(self.INDENT * 9 + "<from>")
+      for fz in options["from_zone"]:
+        rules.append(self.INDENT * 10 + "<member>" + fz + "</member>")
+      rules.append(self.INDENT * 9 + "</from>")
+
+      rules.append(self.INDENT * 9 + "<source>")
+      if options["source"] == []:
```

Comment:
You should use an implicit evaluation
please see https://google.github.io/styleguide/pyguide.html?showone=True/False_evaluations#True/False_evaluations
and from https://www.python.org/dev/peps/pep-0008/
For sequences, (strings, lists, tuples), use the fact that empty sequences are false.

Yes: if not seq:
     if seq:

No: if len(seq):
    if not len(seq):